### PR TITLE
Add select_biased! macro

### PR DESCRIPTION
**Pull request by ryoqun**
~(this is a draft; there's too much of code duplication...)~

much like https://github.com/rust-lang/futures-rs/pull/1976, i want biased select in `crossbeam-channel` as well. :)

### CODE REVIEW

**Comment by ryoqun** [comment]
@taiki-e 

thanks for maintaining this crate. :)

i'll group my gentle reminders with this single comment. but, I'd want some initial feedback for following draft prs. if there's some hope for merging them, I'll finish off these to cross the line, very happily:

- https://github.com/crossbeam-rs/crossbeam/pull/959
- https://github.com/crossbeam-rs/crossbeam/pull/1039
- https://github.com/crossbeam-rs/crossbeam/pull/1040 (this pr)

**Comment by ryoqun** [comment]
@taiki-e hey, thanks for active crossbeam maintenance. could you give this pr a initial assessment whether this pr can be merge or not ever? like other my prs, I'd like all of them to be merged eventually in some way...

thanks in advance.

**Comment by taiki-e** [comment]
Sorry for the slow response, I thought this was a patch that was 1k lines of complexity with no any testing or documentation.

I would like to merge this once the tests and documentation indicating that this is properly biased have been added, and eliminated duplicate code by something like passing a bool literal to an internal macro.

**Comment by ryoqun** [comment]
@taiki-e 

> Sorry for the slow response

no problem. i really thanks for replying!

> I thought this was a patch that was 1k lines of complex implementation without any tests or documentation.

sorry for confusion... I meant this is just pre-rfc api addition with copy-paste impl by marking it as _draft pr_... but i lacked explicit explanation...

> I would like to merge this once the tests and documentation indicating that this is properly biased have been added, and eliminated duplicate code by something like passing a bool literal to an internal macro.

super cool!. I just done all the above after rebased.

File: crossbeam-channel/tests/select_macro.rs
Comment (ryoqun) [inline crossbeam-channel/tests/select_macro.rs:947]: i omitted more test cases like send one. let me know if i need to add more...

File: crossbeam-channel/src/select_macro.rs
Comment (ryoqun) [inline crossbeam-channel/src/select_macro.rs:1147]: i think this was easiest way to implement... but this is accessible to user code. however, it seems we have precedent: `_LEN`.

**Comment by ryoqun** [comment]
> I've just done all the above after rebased. Also marked it as ready for review. :) Thanks in advance.

sorry, i pushed more to fix bugs since this comment...

**Review by taiki-e — APPROVED** [review]
LGTM, thanks @ryoqun!

**Comment by ryoqun** [comment]
@taiki-e Really and really thanks for approving this pr. Then, is this pr in a kind of internal testing now before merging? Also, do i need to add entries to `crossbeam-channel/CHANGELOG.md` via separate pr? I'm not sure the release process of `crossbeam-channel`... Anyway, I'm very looking forward to seeing my first pr for the crossbeam project to be shipped. :)

**Comment by taiki-e** [comment]
> is this pr in a kind of internal testing now before merging

CI is failing on the master branch (mostly due to https://github.com/rust-lang/rust/issues/124800) and we need to fix that before merging PRs...

**Comment by ryoqun** [comment]
> > is this pr in a kind of internal testing now before merging
> 
> CI is failing on the master branch (mostly due to [rust-lang/rust#124800](https://github.com/rust-lang/rust/issues/124800)) and we need to fix that before merging PRs... https://github.com/crossbeam-rs/crossbeam/actions/runs/9072283469

Thanks for the quick reply. hmm, that's quite unfortunate, considering this pr is so close to be shipped... I just skimmed over the rust issue and its background. That said, I wonder how soon this resolves upstream..

Maybe pinning to older nightly (nightly-2024-05-04 or earlier according to https://blog.rust-lang.org/2024/05/06/check-cfg.html) in crossbeam CI isn't a viable option?

**Comment by ryoqun** [comment]
> we need to fix that before merging PRs...
> https://github.com/crossbeam-rs/crossbeam/actions/runs/9072283469

@taiki-e fyi, i fixed that: https://github.com/crossbeam-rs/crossbeam/pull/1109 :)

**Comment by ryoqun** [comment]
@taiki-e I'm really appreciate merging this pr (and the ci fix pr as well). Now, the only thing I'm looking forward to is `crossbeam-channel 0.5.13`... sorry for asking too much :)

**Comment by taiki-e** [comment]
Published in crossbeam-channel 0.5.13.

**Comment by ryoqun** [comment]
@taiki-e hi, I'm reporting back with a good news. :) Really thanks to your collation with this pr shipped, using `select_biased!` resulted in 12-13% performance improvement: https://github.com/anza-xyz/agave/pull/1437. and the pr just got merged there.

**Pull request merged**
